### PR TITLE
LocalExecutor rewritten to use a local server instead of creating a n…

### PR DESCRIPTION
**Pull request by dimberman**
…ew process for every task

Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/executors/local_executor.py [hunk 1]
```diff
@@ -46,185 +43,136 @@
...
+from asyncio.futures import Future
+from functools import partial
+from requests import Response
+import aiohttp
+
+
+async def make_request_async(task_id, dag_id, execution_date) -> aiohttp.ClientResponse:
+    # req = 'http://35.245.62.83/run'
+    req = "http://localhost:8081/run"
+    date = int(datetime.datetime.timestamp(execution_date))
```
Comment (ashb) [inline airflow/executors/local_executor.py:59]: This truncates to seconds/drops milliseconds which it shouldn't (for manually triggered dags the e_d can include subseconds)

File: airflow/executors/local_executor.py [hunk 2]
```diff
@@ -46,185 +43,136 @@
...
+    params = {
+        "task_id": task_id,
+        "dag_id": dag_id,
+        "execution_date": date,
+        # "subdir": "/root/airflow/dags "
+    }
+
+    async with aiohttp.ClientSession() as session:
+        # async with session.get(url=req, params=params, headers={"Host": "airflow-knative.default.example.com"}) as resp:
+        async with session.get(url=req, params=params, ) as resp:
```
Comment (ashb) [inline airflow/executors/local_executor.py:67]: This should probably be a POST request - GET's shouldn't "modify state"

File: airflow/executors/local_executor.py [hunk 3]
```diff
@@ -46,185 +43,136 @@
...
         """
         if key is None:
             return
-        self.log.info("%s running %s", self.__class__.__name__, command)
+        self.log.info("%s running %s", self.__class__.__name__, task_instance)
         try:
-            subprocess.check_call(command, close_fds=True)
-            state = State.SUCCESS
-        except subprocess.CalledProcessError as e:
+            date = int(datetime.datetime.timestamp(task_instance.execution_date))
```
Comment (ashb) [inline airflow/executors/local_executor.py]: unused here? duplicate of make_request_async anyway

File: airflow/executors/local_executor.py [hunk 4]
```diff
@@ -46,185 +43,136 @@
...
-        self.impl.execute_async(key=key, command=command)
+    def execute_group_async(self,
+                            task_instances: multiprocessing.Queue = None):
+        tasks_to_run = []
+        tasks = []
+        while not task_instances.empty():
+            tasks_to_run.append(task_instances.get())
+        for t in tasks_to_run:
+            (key, ti) = t
+            make_request_async(ti.task_id, ti.dag_id, ti.execution_date)
```
Comment (ashb) [inline airflow/executors/local_executor.py]: Doesn't this _also_ make a request, in addition to the next line?

File: airflow/worker/airflow_worker.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    dag_id = request.rel_url.query['dag_id']
+    task_id = request.rel_url.query['task_id']
+    # subdir = request.rel_url.query['subdir']
+    subdir = "/Users/dimberman/airflow/dags"
+
+    execution_date = pendulum.fromtimestamp(int(request.rel_url.query["execution_date"]))
+    log = LoggingMixin().log
+    #
+    log.info("running dag {} for task {} on date {} in subdir {}".format(dag_id, task_id, execution_date, subdir))
+    logging.shutdown()
```
Comment (ashb) [inline airflow/worker/airflow_worker.py:82]: What effect does having this here have on the logs from the task? Does it get re-initiallized in the task runner?

File: airflow/bin/cli.py [hunk 6]
```diff
@@ -1042,6 +1042,47 @@ def monitor_gunicorn(gunicorn_master_proc):
             monitor_gunicorn(gunicorn_master_proc)
 
 
+@cli_utils.action_logging
+def airflow_worker(args):
```
Comment (schnie) [inline airflow/bin/cli.py:1090]: @dimberman so this means when we fire up a knative service our command will be `airflow airflow_worker`? And celery workers are spun up now with `airflow worker`. Feels a little confusing. Not sure of anything better yet :thinking:

**Pull request closed without merging**